src/main/java/com/iota/iri/network/NeighborRouter.java

```diff
@@ -0,0 +1,840 @@
+                neighborURI.getPort()));
+        reconnectPool.remove(rawURI);
+
+        String identity = String.format("%s:%d", inetAddr.getAddress().getHostAddress(), inetAddr.getPort());
+        Neighbor neighbor = connectedNeighbors.get(identity);
+        if (neighbor == null) {
+            return NeighborMutOp.UNKNOWN_NEIGHBOR;
+        }
+
+        // the neighbor will be disconnected inside the selector loop
+        neighbor.setState(NeighborState.MARKED_FOR_DISCONNECT);
+        return NeighborMutOp.OK;
+    }
+
+    /**
+     * Parses the given string to an URI. The URI must use "tcp://" as the protocol.
+     *
+     * @param uri The URI string to parse
+     * @return the parsed URI, if parsed correctly
+     */
+    public static Optional<URI> parseURI(final String uri) {
+        if (uri.isEmpty()) {
+            return Optional.empty();
+        }
+
+        URI neighborURI;
+        try {
+            neighborURI = new URI(uri);
+        } catch (URISyntaxException e) {
+            log.error("URI {} raised URI Syntax Exception", uri);
+            return Optional.empty();
+        }
+        if (!isURIValid(neighborURI)) {
+            return Optional.empty();
+        }
+        return Optional.of(neighborURI);
+    }
+
+    /**
+     * Checks whether the given URI is valid. The URI is valid if it is not null and it uses TCP as the protocol.
+     * 
+     * @param uri The URI to check
+     * @return true if the URI is valid, false if not
+     */
+    public static boolean isURIValid(final URI uri) {
+        if (uri == null) {
+            log.error("Cannot read URI schema, please check neighbor config!");
+            return false;
+        }
+
+        if (!uri.getScheme().equals("tcp")) {
+            log.error("'{}' is not a valid URI schema", uri);
+            return false;
+        }
+        return true;
+    }
+
+    /**
+     * Returns the {@link TransactionProcessingPipeline}.
+     * 
+     * @return the {@link TransactionProcessingPipeline} used by the {@link NeighborRouter}
+     */
+    public TransactionProcessingPipeline getTransactionProcessingPipeline() {
+        return txPipeline;
+    }
+
+    /**
+     * Gets all neighbors the {@link NeighborRouter} currently sees as either connected or attempts to build connections
+     * to.
+     * 
+     * @return The neighbors
+     */
+    public List<Neighbor> getNeighbors() {
+        List<Neighbor> neighbors = new ArrayList<>(connectedNeighbors.values());
+        reconnectPool.forEach(uri -> {
+            neighbors.add(new NeighborImpl<>(null, null, uri.getHost(), uri.getPort(), null));
+        });
+        return neighbors;
+    }
+
+    /**
+     * Gets the currently connected neighbors.
+     * 
+     * @return The connected neighbors.
+     */
+    public Map<String, Neighbor> getConnectedNeighbors() {
+        return connectedNeighbors;
```

Comment:
We don't want other threads to be modifying this map.
We have 2 options (that I can think of at least):
1. `return Collections.unmodifiableMap()` - this will create a wrapper around the original map. All changes to the underlying map will be reflected in the new map. The disadvantage is that you are creating more objects (they are just wrappers so it is not so bad, however this will add gc pressure if the method is called a lot)

2. Change the return type to `Iterable<Neighbor>` and return the `connectedNeighbors.values`. This will not create additional objects. However, it doesn't provide complete protection since you can still remove objects with an iterator (you are only protected against addition). You will also need another method to expose the size

Anyhows, since I doubt that the bottleneck of IRI will ever be because too many `unmodifiable` wrappers are created I will opt for the first.

Reply:
I don't think returning an immutable map is the right thing to do here. `getConnectedNeighbors()` is called in very tight loops (see BroadcastStage) so a lot of garbage would be created by doing something like this. We have 100% knowledge where the map is being used, so I don't see the point of protecting it from getting modified.

However, I am going to retuirn an unmodifiable map.